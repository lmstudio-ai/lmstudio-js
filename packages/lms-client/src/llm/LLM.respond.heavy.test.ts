import { Chat, ChatMessage, type LLM, LMStudioClient } from "../index.js";
import { ensureHeavyTestsEnvironment, llmTestingModel } from "../shared.heavy.test.js";

describe("LLM", () => {
  let client: LMStudioClient;
  let model: LLM;
  const chat = Chat.from([
    { role: "system", content: "This is the system prompt." },
    { role: "user", content: "User message 1" },
    { role: "assistant", content: "Assistant message 1" },
    { role: "user", content: "User message 2" },
  ]);
  beforeAll(async () => {
    client = new LMStudioClient();
    await ensureHeavyTestsEnvironment(client);
  });
  beforeEach(async () => {
    model = await client.llm.model(llmTestingModel, { verbose: false });
  });
  describe(".respond", () => {
    it("should work without streaming", async () => {
      const result = await model.respond(chat, {
        temperature: 0,
        maxTokens: 10,
      });
      expect(result.content).toMatchInlineSnapshot(`"Assistant message 2"`);
      expect(result.stats).toMatchSnapshot({
        numGpuLayers: expect.any(Number),
        timeToFirstTokenSec: expect.any(Number),
        tokensPerSecond: expect.any(Number),
      });
      expect(result.modelInfo).toMatchSnapshot({
        identifier: expect.any(String),
        instanceReference: expect.any(String),
        modelKey: expect.any(String),
      });
      expect(result.roundIndex).toEqual(0);
    });
    it("should work with streaming", async () => {
      const prediction = model.respond(chat, {
        temperature: 0,
        maxTokens: 10,
      });
      const fragments = [];
      for await (const fragment of prediction) {
        fragments.push(fragment);
      }
      expect(fragments).toMatchSnapshot();
      const result = await prediction.result();
      expect(result.content).toMatchSnapshot();
      expect(result.stats).toMatchSnapshot({
        numGpuLayers: expect.any(Number),
        timeToFirstTokenSec: expect.any(Number),
        tokensPerSecond: expect.any(Number),
      });
      expect(result.modelInfo).toMatchSnapshot({
        identifier: expect.any(String),
        instanceReference: expect.any(String),
        modelKey: expect.any(String),
      });
    });
    it("should allow cancel via .cancel()", async () => {
      const prediction = model.respond(chat, {
        temperature: 0,
        maxTokens: 50,
      });
      prediction.cancel();
      const result = await prediction.result();
      expect(result.stats.stopReason).toEqual("userStopped");
    });
    it("should allow cancel via abort controller", async () => {
      const controller = new AbortController();
      const prediction = model.respond(chat, {
        temperature: 0,
        maxTokens: 50,
        signal: controller.signal,
      });
      controller.abort();
      const result = await prediction.result();
      expect(result.stats.stopReason).toEqual("userStopped");
    });
    it("should call onFirstToken callback", async () => {
      const onFirstToken = jest.fn();
      await model.respond(chat, {
        temperature: 0,
        maxTokens: 1,
        onFirstToken,
      });
      expect(onFirstToken).toHaveBeenCalled();
    });
    it("should call onPromptProcessingProgress callback", async () => {
      const onPromptProcessingProgress = jest.fn();
      await model.respond(chat, {
        temperature: 0,
        maxTokens: 1,
        onPromptProcessingProgress,
      });
      expect(onPromptProcessingProgress).toHaveBeenCalledWith(0);
      expect(onPromptProcessingProgress).toHaveBeenCalledWith(1);
    });
    it("should call onPredictionFragment callback", async () => {
      const onPredictionFragment = jest.fn();
      await model.respond(chat, {
        temperature: 0,
        maxTokens: 10,
        onPredictionFragment,
      });
      const calls = onPredictionFragment.mock.calls;
      expect(calls).toMatchSnapshot();
    });
    it("should call onMessage callback", async () => {
      const onMessage = jest.fn();
      await model.respond(chat, {
        temperature: 0,
        maxTokens: 10,
        onMessage,
      });
      expect(onMessage).toHaveBeenCalledTimes(1);
      const call = onMessage.mock.calls[0];
      const message = call[0] as ChatMessage;
      expect(message).toBeInstanceOf(ChatMessage);
      expect(message.getRole()).toEqual("assistant");
      expect(message.getText()).toEqual("Assistant message 2");
    });
    it("should work with single string input", async () => {
      const result = await model.respond("User message 1", {
        temperature: 0,
        maxTokens: 10,
      });
      expect(result.content).toMatchInlineSnapshot(`"Hello! How can I assist you today?"`);
      expect(result.stats).toMatchSnapshot({
        numGpuLayers: expect.any(Number),
        timeToFirstTokenSec: expect.any(Number),
        tokensPerSecond: expect.any(Number),
      });
      expect(result.modelInfo).toMatchSnapshot({
        identifier: expect.any(String),
        instanceReference: expect.any(String),
        modelKey: expect.any(String),
      });
      expect(result.roundIndex).toEqual(0);
    });
    it("should work with array of messages input", async () => {
      const result = await model.respond(
        [
          { role: "system", content: "This is the system prompt." },
          { role: "user", content: "User message 1" },
          { role: "assistant", content: "Assistant message 1" },
          { role: "user", content: "User message 2" },
        ],
        {
          temperature: 0,
          maxTokens: 10,
        },
      );
      expect(result.content).toMatchInlineSnapshot(`"Assistant message 2"`);
      expect(result.stats).toMatchSnapshot({
        numGpuLayers: expect.any(Number),
        timeToFirstTokenSec: expect.any(Number),
        tokensPerSecond: expect.any(Number),
      });
      expect(result.modelInfo).toMatchSnapshot({
        identifier: expect.any(String),
        instanceReference: expect.any(String),
        modelKey: expect.any(String),
      });
      expect(result.roundIndex).toEqual(0);
    });
  });
});
