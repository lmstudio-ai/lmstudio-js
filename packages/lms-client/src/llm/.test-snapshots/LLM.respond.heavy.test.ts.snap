// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LLM .respond should call onPredictionFragment callback 1`] = `
[
  [
    {
      "containsDrafted": false,
      "content": "Assistant",
      "reasoningType": "none",
      "tokensCount": 1,
    },
  ],
  [
    {
      "containsDrafted": false,
      "content": " message",
      "reasoningType": "none",
      "tokensCount": 1,
    },
  ],
  [
    {
      "containsDrafted": false,
      "content": " ",
      "reasoningType": "none",
      "tokensCount": 1,
    },
  ],
  [
    {
      "containsDrafted": false,
      "content": "2",
      "reasoningType": "none",
      "tokensCount": 1,
    },
  ],
]
`;

exports[`LLM .respond should work with array of messages input 2`] = `
{
  "numGpuLayers": Any<Number>,
  "predictedTokensCount": 5,
  "promptTokensCount": 41,
  "stopReason": "eosFound",
  "timeToFirstTokenSec": Any<Number>,
  "tokensPerSecond": Any<Number>,
  "totalTokensCount": 46,
}
`;

exports[`LLM .respond should work with array of messages input 3`] = `
{
  "architecture": "qwen2",
  "contextLength": 4096,
  "displayName": "Qwen2.5 0.5B Instruct",
  "format": "gguf",
  "identifier": Any<String>,
  "instanceReference": Any<String>,
  "maxContextLength": 32768,
  "modelKey": Any<String>,
  "paramsString": "1B",
  "path": "lmstudio-community/Qwen2.5-0.5B-Instruct-GGUF/Qwen2.5-0.5B-Instruct-Q4_K_M.gguf",
  "sizeBytes": 397807936,
  "trainedForToolUse": true,
  "type": "llm",
  "vision": false,
}
`;

exports[`LLM .respond should work with single string input 2`] = `
{
  "numGpuLayers": Any<Number>,
  "predictedTokensCount": 10,
  "promptTokensCount": 33,
  "stopReason": "maxPredictedTokensReached",
  "timeToFirstTokenSec": Any<Number>,
  "tokensPerSecond": Any<Number>,
  "totalTokensCount": 43,
}
`;

exports[`LLM .respond should work with single string input 3`] = `
{
  "architecture": "qwen2",
  "contextLength": 4096,
  "displayName": "Qwen2.5 0.5B Instruct",
  "format": "gguf",
  "identifier": Any<String>,
  "instanceReference": Any<String>,
  "maxContextLength": 32768,
  "modelKey": Any<String>,
  "paramsString": "1B",
  "path": "lmstudio-community/Qwen2.5-0.5B-Instruct-GGUF/Qwen2.5-0.5B-Instruct-Q4_K_M.gguf",
  "sizeBytes": 397807936,
  "trainedForToolUse": true,
  "type": "llm",
  "vision": false,
}
`;

exports[`LLM .respond should work with streaming 1`] = `
[
  {
    "containsDrafted": false,
    "content": "Assistant",
    "reasoningType": "none",
    "tokensCount": 1,
  },
  {
    "containsDrafted": false,
    "content": " message",
    "reasoningType": "none",
    "tokensCount": 1,
  },
  {
    "containsDrafted": false,
    "content": " ",
    "reasoningType": "none",
    "tokensCount": 1,
  },
  {
    "containsDrafted": false,
    "content": "2",
    "reasoningType": "none",
    "tokensCount": 1,
  },
]
`;

exports[`LLM .respond should work with streaming 2`] = `"Assistant message 2"`;

exports[`LLM .respond should work with streaming 3`] = `
{
  "numGpuLayers": Any<Number>,
  "predictedTokensCount": 5,
  "promptTokensCount": 41,
  "stopReason": "eosFound",
  "timeToFirstTokenSec": Any<Number>,
  "tokensPerSecond": Any<Number>,
  "totalTokensCount": 46,
}
`;

exports[`LLM .respond should work with streaming 4`] = `
{
  "architecture": "qwen2",
  "contextLength": 4096,
  "displayName": "Qwen2.5 0.5B Instruct",
  "format": "gguf",
  "identifier": Any<String>,
  "instanceReference": Any<String>,
  "maxContextLength": 32768,
  "modelKey": Any<String>,
  "paramsString": "1B",
  "path": "lmstudio-community/Qwen2.5-0.5B-Instruct-GGUF/Qwen2.5-0.5B-Instruct-Q4_K_M.gguf",
  "sizeBytes": 397807936,
  "trainedForToolUse": true,
  "type": "llm",
  "vision": false,
}
`;

exports[`LLM .respond should work without streaming 2`] = `
{
  "numGpuLayers": Any<Number>,
  "predictedTokensCount": 5,
  "promptTokensCount": 41,
  "stopReason": "eosFound",
  "timeToFirstTokenSec": Any<Number>,
  "tokensPerSecond": Any<Number>,
  "totalTokensCount": 46,
}
`;

exports[`LLM .respond should work without streaming 3`] = `
{
  "architecture": "qwen2",
  "contextLength": 4096,
  "displayName": "Qwen2.5 0.5B Instruct",
  "format": "gguf",
  "identifier": Any<String>,
  "instanceReference": Any<String>,
  "maxContextLength": 32768,
  "modelKey": Any<String>,
  "paramsString": "1B",
  "path": "lmstudio-community/Qwen2.5-0.5B-Instruct-GGUF/Qwen2.5-0.5B-Instruct-Q4_K_M.gguf",
  "sizeBytes": 397807936,
  "trainedForToolUse": true,
  "type": "llm",
  "vision": false,
}
`;
